Modules/Devices/src/Devices.Infrastructure/Persistence/Repository/OAuthClientsRepository.cs

```diff
@@ -13,21 +14,38 @@ namespace Backbone.Modules.Devices.Infrastructure.Persistence.Repository;
 public class OAuthClientsRepository : IOAuthClientsRepository
 {
     private readonly OpenIddictApplicationManager<CustomOpenIddictEntityFrameworkCoreApplication> _applicationManager;
+    private List<CustomOpenIddictEntityFrameworkCoreApplication> _cachedApplications;
 
     public OAuthClientsRepository(OpenIddictApplicationManager<CustomOpenIddictEntityFrameworkCoreApplication> applicationManager)
     {
         _applicationManager = applicationManager;
+        _cachedApplications = new List<CustomOpenIddictEntityFrameworkCoreApplication>();
     }
 
     public async Task<IEnumerable<OAuthClient>> FindAll(CancellationToken cancellationToken)
     {
-        var clients = await _applicationManager.ListAsync(applications => applications.Select(c => new OAuthClient(c.ClientId, c.DisplayName, c.TierId)), cancellationToken).ToListAsync(cancellationToken);
-        return clients;
+        var applications = await _applicationManager.ListAsync(cancellationToken: cancellationToken).ToListAsync(cancellationToken);
+
+        _cachedApplications = applications;
+
+        return applications.Select(client => new OAuthClient(client.ClientId!, client.DisplayName!, client.DefaultTier)).ToList();
     }
 
-    public async Task<CustomOpenIddictEntityFrameworkCoreApplication> Find(string clientId, CancellationToken cancellationToken)
+    public async Task<OAuthClient> Find(string clientId, CancellationToken cancellationToken, bool track = false)
     {
-        return await _applicationManager.FindByClientIdAsync(clientId, cancellationToken) ?? throw new NotFoundException(nameof(OAuthClient));
+        var application = await _applicationManager.FindByClientIdAsync(clientId, cancellationToken) ?? throw new NotFoundException(nameof(OAuthClient));
+
+        var cachedApplicationIndex = _cachedApplications.FindIndex(cachedApplication => cachedApplication.ClientId == clientId);
+        if (cachedApplicationIndex != -1)
+        {
+            _cachedApplications[cachedApplicationIndex] = application;
+        }
+        else
+        {
+            _cachedApplications.Add(application);
+        }
+
+        return new OAuthClient(application.ClientId!, application.DisplayName!, application.DefaultTier);
```

Comment:
You shouldn't update the tracked entity. Once it is tracked. this entity should always be used. Also, you should check the `track` paramter.

```suggestion
        if(_trackedApplications.ContainsKey(clientId)) {
            return _trackedApplications[clientId];
        
        var application = await _applicationManager.FindByClientIdAsync(clientId, cancellationToken);
        
        if(track)
            _trackedApplications[clientId]= application;

        return new OAuthClient(application.ClientId!, application.DisplayName!, application.DefaultTier);
```